aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu

```diff
@@ -708,4 +714,290 @@ Tensor _sparse_sum_backward_cuda(const Tensor& grad_, const SparseTensor& input_
+  return _bmm_out_sparse_cuda(result, self, mat2, false);
+}
+
+Tensor& _bmm_out_sparse_cuda(Tensor& result, const SparseTensor& self, const Tensor& mat2, bool deterministic) {
+#ifdef __HIP_PLATFORM_HCC__
+  TORCH_CHECK(false, "bmm sparse-dense is not supported on HIP")
+#else
+
+  TORCH_CHECK(!mat2.is_sparse(), "bmm_sparse: Tensor 'mat2' must be dense");
+  TORCH_CHECK(self.dense_dim() == 0, "bmm_sparse: Tensor 'self' must have 0 dense dims, but has ", self.dense_dim());
+  TORCH_CHECK(self.sparse_dim() == 3, "bmm_sparse: Tensor 'self' must have 3 sparse dims, but has ", self.sparse_dim());
+  TORCH_CHECK(mat2.dim() == 3, "bmm_sparse: Tensor 'mat2' must have 3 dims, but has ", mat2.dim());
+  TORCH_CHECK(self.size(0) == mat2.size(0), "bmm_sparse: 'self.size(0)' and 'mat2.size(0)' must match");
+  TORCH_CHECK(self.size(2) == mat2.size(1), "bmm_sparse: 'self.size(2)' and 'mat2.size(1)' must match");
+
+  int64_t num_matrices = self.size(0);
+  int64_t dim_i = self.size(1);
+  int64_t dim_j = self.size(2);
+  int64_t dim_k = mat2.size(2);
+
+  result.resize_({num_matrices, dim_k, dim_i});
+
+  if ((self._nnz() == 0) || (dim_j == 0) || (dim_k == 0)) {
+    result.zero_().transpose_(1, 2);
+    return result;
+  }
+
+  Tensor tmp_result;
+  bool need_copy_result;
+
+  // If the result tensor is contiguous, we can just write results directly to it.
+  // Otherwise, we'll need to write results to a temp buffer and then copy.
+  if (result.is_contiguous()) {
+    tmp_result = result;
+    need_copy_result = false;
+  } else {
+    tmp_result = at::empty({num_matrices, dim_k, dim_i}, result.options(), at::MemoryFormat::Contiguous);
+    need_copy_result = true;
+  }
+
+  // Dense matrices have to be contiguous for cusparseSpMM to work
+  const Tensor mat2_contig = mat2.is_contiguous() ? mat2 : mat2.clone(at::MemoryFormat::Contiguous);
+  auto cusparse_handle = at::cuda::getCurrentCUDASparseHandle();
+
+  // First need to coalesce to get all of the first dimension indices
+  // in order since we'll be sending each matrix into the MM operation
+  SparseTensor self_coalesced = coalesce_sparse_cuda(self);
+
+  int64_t nnz =        self_coalesced._nnz();
+  LongTensor indices = self_coalesced._indices();
+  Tensor values =      self_coalesced._values();
+
+  LongTensor indices_dim0 = indices[0];
+
+  // Need to convert dim1 and dim2 indices to 32-bit since cusparseSpMM
+  // only supports 32-bit indices
+  Tensor indices_dim1 = indices[1].to(ScalarType::Int);
+  Tensor indices_dim2 = indices[2].to(ScalarType::Int);
+
+  int64_t* mat_el_end_indices;
+  cudaMallocManaged(&mat_el_end_indices, num_matrices*sizeof(int64_t));
+  search_end_matrix_indices(mat_el_end_indices, num_matrices, indices_dim0);
+
+  Scalar beta = 0;
+  Scalar alpha = 1;
+
+  int64_t mat_el_begin_idx = 0;
+  size_t* workspace_buffer_sizes = new size_t[num_matrices];
+  void** workspace_buffers = new void*[num_matrices];
```

Comment:
You sure there would not be any errors in-between to cause memory leaks?

Reply:
Good point, if an exception is thrown, then the deletions at the end of the function wouldn't get run. I could put the array of buffers on the stack instead. But the buffers themselves have to be freed with cudaFree(), so I guess I need to use a try/catch to handle that case.

Reply:
Maybe I am wrong, but I am not sure you actually allocate this memory on gpu...

Reply:
Yep `std::vector` is probably what I'll use.

Reply:
Is the reason you need `num_matrices` workspace buffers because you run all of the individual unbatched sparse matrix multiplies in parallel?

Reply:
Not exactly. From what I understand, each cuda function call gets sent to the gpu serially (correct me if I'm wrong). Avoiding the cudaFree between each matrix operation gave a measurable speedup. I'm not 100% sure why that is, and I don't recall how much speedup.

Reply:
You're allocating and deallocating cuda manually; that's REALLY slow. It would be much better to get the scratch buffers by allocating a regular old tensor and then the memory will come from our caching allocator. But isn't the buffer the same size every call? So I don't think you need to allocate one buffer for everything.

Reply:
The buffer might not be the same size each time, I'm not sure. It could potentially be different if the cusparseSpMM algorithm needs less space for fewer non-zero elements, for instance. But it's probably a better approach to reuse the same buffer and just increase its size only when needed.

Reply:
Yes. In particular, this will reduce the maximum memory residency since you don't preallocate every buffer.

Reply:
Done